cabalscan/src/CabalScan/RuleGenerator.hs

```diff
@@ -223,18 +227,38 @@ generateRule cabalFilePath pkgId dataFiles bi someModules ctype attrName pubAttr
     toToolName (Cabal.ExeDependency pkg exe _) =
       ToolName (pkgNameToText pkg) (Text.pack $ Cabal.unUnqualComponentName exe)
 
-mainFileAttr :: Path b File -> FilePath -> Cabal.BuildInfo -> IO Attributes
-mainFileAttr cabalFilePath mainFilePath bi = do
-  let
-    hsSourceDirs = Cabal.hsSourceDirs bi
-    cabalParentDir = Path.parent cabalFilePath
-    filePathToAttrValue = TextValue . Text.pack . Path.toFilePath
-    pathToMainFileAttr :: Path Rel File -> (Text, AttrValue)
-    pathToMainFileAttr = ((,) "main_file") . filePathToAttrValue
-  mainModulePath <- Path.parseRelFile $ dropExtension mainFilePath
-  srcDirs <- mapM Path.parseRelDir hsSourceDirs
-  pathMaybe <- findModulePath cabalParentDir srcDirs mainModulePath
-  return $ maybeToList $ fmap pathToMainFileAttr pathMaybe
+-- | Thrown when we can't find the file path of a main
+-- file which is referenced in a Cabal file (hs-source-dirs + main-is).
+data MainFileNotFound = MainFileNotFound
+  { cabalFile :: FilePath
+  , mainFile :: FilePath
+  , hsSourceDirs :: [FilePath]
+  }
+  deriving (Show, Exception)
+
+mainFileAttr :: FilePath -> (Text, AttrValue)
+mainFileAttr mainPath = ("main_file", TextValue $ Text.pack mainPath)
+
+-- | @findMainFile cabalFile mainFile hsSrcDirs@
+--
+-- Finds out relative path to the main file under the directory where
+-- the Cabal file is. It does simple concatenation between @hsSrcDir@ and
+-- the @mainFile@ and validates the existence of such path.
+--
+findMainFile :: Path b File -> FilePath -> [FilePath] -> IO FilePath
+findMainFile cabalFile mainFile hsSrcDirs = do
+  let parentDir = Path.parent cabalFile
+  mainPath <- Path.parseRelFile mainFile
+  srcDirs <- mapM Path.parseRelDir hsSrcDirs
+  let mainPaths = [ dir Path.</> mainPath | dir <- srcDirs ]
+  validPaths <- filterM (Path.doesFileExist . ((Path.</>) parentDir)) mainPaths
+  case validPaths of
+    path:_ -> return $ Path.toFilePath path
```

Comment:
Should we fail if there are multiple files found? This could eliminate confusion by accidentally picking an unexpected file.